### ISSUE TITLE
Protocol: allow `load_profile` to detect when a "development" version is being used

### ISSUE THREAD

**Issue by sphuber** [aiidateam/aiida-core#5290]
## Problem description
The current policy is that the code on `develop` is not officially released and therefore not officially supported. Concretely this means that we reserve the right to not only put backwards-incompatible changes with respect to the interface, but also to freely change migrations that not have yet been released yet. The latter poses potential problems if people use their database with the develop branch. If we change migrations they have already applied, if they at some point use the database with the next released version, the migrations may fail, or they may not actually include the correct migrations (since there database has applied the incorrect ones).

Even though we do not officially support this usage and the migrate command makes it very clear to make a backup, there will be users who do not realize this and end up in a situation where their production database may be incompatible with the next official release.

One solution is to never modify migrations once merged, but always to create additional migrations to correct the mistake. Putting the considerable amount of additional work that this will require, in certain cases it is also not always possible, because the incorrect migration may have been lossy and so a follow-up migration would be incapable of restoring the lost data.

Therefore, the only solution is to make it way more clear to users when they are using a non-released version of the code.

## Proposed solution
We adopt a new policy where the version specified by the version identifier `aiida.__version__` always includes the `-dev` suffix on the `develop` branch. The `load_profile` method will include a check to see if the current installed code has a version with this suffix and will emit a warning. We should probably also change the default branch to be `master` instead of `develop`, such that if users clone the repo and install it, they would at least install a released version.

For developers, the printed warning may be very annoying, so we should probably add a configuration option to silence it.

The next release planned is `2.0` and so we could introduce this new protocol then.

**Comment by sphuber** [comment]
@giovannipizzi @ltalirz @chrisjsewell @ramirezfranciscof

**Comment by giovannipizzi** [comment]
I fully agree that this is very much needed if we want to apply the approach of fixing migrations in develop before a release (for which I see the need, in terms of human development costs). Without a clear warning I think we cannot assume a user is aware of the fact they shouldn't use the version from `git`, or we risk to have many disappointed users that weren't aware of our assumptions on which versions are usable, and just stop using AiiDA.

I think that changing the default branch might not even be needed if the warning is clear enough.

I'm not sure how annoying the banner will be, but if this turns out to be a problem, yes, we can find a config option for this; I would wait to see if this is truly needed, though (and we shouldn't advertise it generally, but only put it in the wiki, to avoid that people turn it immediately off as the first thing, defeating the purpose of the banner).

So in summary I strongly support this.

**Comment by ltalirz** [comment]
This is very similar to https://github.com/aiidateam/aiida-core/issues/4551 , so I agree with the need.

As mentioned there, we can also adopt a solution (versioneer/ ...) that others have been using instead of rolling our own, but up to you guys.

**Comment by jbweston** [comment]
I commented on https://github.com/aiidateam/aiida-core/issues/4551#issuecomment-1003441929, there is https://github.com/jbweston/miniver/.

It does the same job as versioneer, but involves including only a `_version.py` file with <200 lines.

It's also public domain, so doesn't require any licensing shenanigans.

**Comment by ltalirz** [comment]
Joe's solution looks good to me

**Comment by sphuber** [comment]
Thanks for the link @jbweston . I was indeed wary of the `versioneer` solution since it adds a big file that will itself have to be maintained. Your solution goes in the right direction by reducing the complexity. Still, I am not sure we need all that. I don't think we need versioning that is more granular than released versions and go to commits. As discussed elsewhere, this can even have undesirable effects with regard to our caching system. In the end, all we need is to add a simple commit that adds a suffix to the version in `aiida/__init__.py` when a new release is made and merged back into develop. I would propose we simply include this in the release protocol. It is simple, effective and doesn't require additional tools that need to be maintained.

**Comment by chrisjsewell** [comment]
I agree @sphuber 👍 

On a semi-related note, flit now supports editable builds: https://flit.readthedocs.io/en/latest/history.html#version-3-4, is now officially part of the python packaging authority: https://github.com/pypa/flit, and already supports PEP 621: https://flit.readthedocs.io/en/latest/pyproject_toml.html#new-style-metadata 👀 
So I'm tempted to suggest we move to that, and get then ditch `setup.json` and `setup.py`, and have everything in `pyproject.toml`.
I know you gotta love that 😄

**Comment by sphuber** [comment]
> On a semi-related note, flit now supports editable builds: https://flit.readthedocs.io/en/latest/history.html#version-3-4, is now officially part of the python packaging authority: https://github.com/pypa/flit, and already supports PEP 621: https://flit.readthedocs.io/en/latest/pyproject_toml.html#new-style-metadata eyes
> So I'm tempted to suggest we move to that, and can then ditch `setup.json` and `setup.py`, and have everything in `pyproject.toml`.
> I know you gotta love that smile

If the currently supported build methods will continue to work, I would be in favor of this, or even have the `setup.cfg` intermediate. Would this work with `pip`, `conda` etc.?

**Comment by chrisjsewell** [comment]
> Would this work with pip, conda etc.?

I already use in https://github.com/aiidateam/archive-path, and yeh now it is literally under the same organisation as pip

> I doubt many users actually use the newer package managers like flit and poetry

I would envisage that flit will eventually become the defacto packaging tool for pure python packages: it is setuptools without the added complexity required to support legacy packages and C extensions etc

**Comment by giovannipizzi** [comment]
If it's easy to do this as part of the release procedure, I vote that for 2.0 we just extend our internal release policies to do it (and of course add the code to show the warning for relevant operations). 
After 2.0, if we realise this is annoying for us as developers, and we need to use external tools, we can always add them and automate a bit more the procedure for the following releases.

Two questions:
- once we release 2.0, we just append a dev suffix to 2.0, or we already upgrade the version to the next one, i.e. 2.1-dev (assuming we know what it is?)
- (I didn't check; and partially related to the previous point) Is `-dev` a standard recognised suffix from libraries in python dealing with the version number like setuptools? and in addition, is 2.0-dev larger or smaller than 2.0? This might be important to check also to answer my previous point (we want the development version to be larger than the previously released version)

**Comment by sphuber** [comment]
There are standards and they are specified by [PEP 440](https://www.python.org/dev/peps/pep-0440/#local-version-identifiers). Following those guidelines, I propose we simply use a "developmental release" specifier. The most standard version would be to add the suffix `.dev0`. It is possible to omit the `0` and one can use `-` or `_` as the separator, but according to PEP440, using `.` and making the `0` explicit is the standard. So after releasing `v2.0.0`, we merge `master` back into `develop` and add a commit that sets
```
__version__ = '2.0.0.dev0'
```

**Comment by chrisjsewell** [comment]
```python
In [1]: from packaging.version import parse
In [2]: parse("2.0.0.dev0").is_devrelease
Out[2]: True
In [3]: parse("2.0.0.dev0").dev
Out[3]: 0
```

👍

**Comment by giovannipizzi** [comment]
Thanks to both.
The problem however is that 
```python
In [1]: from packaging.version import parse
In [2]: v2 = parse("2.0.0")
In [3]: v2dev = parse("2.0.0.dev0")
In [4]: v2 < v2dev
Out[4]: False
In [5]: v2 > v2dev
Out[5]: True
```
i.e. `2.0.0.dev0` represents the first development version leading to 2.0.0 (what we have now, in essence) while we want the opposite.

A possible solution would be to use `2.0.0.post0.dev0` (or there are other better ideas?)
```python
In [1]: from packaging.version import parse
In [2]: v2 = parse("2.0.0")
In [3]: v2dev = parse('2.0.0.post0.dev0')
In [4]: v2 < v2dev
Out[4]: True
In [5]: v2 > v2dev
Out[5]: False
In [6]: v2dev.is_devrelease
Out[6]: True
In [7]: v2dev.dev
Out[7]: 0
```

**Comment by sphuber** [comment]
I misunderstood then, I thought the develop segment was to indicate a version after release, but apparently not. Then we should simply use `2.0.0.post0`. There is no need to use both segment specifiers:
```python
In [1]: from packaging.version import parse

In [2]: v1 = parse('2.0.0')

In [3]: v2 = parse('2.0.0.post0')

In [4]: v1 > v2
Out[4]: False

In [5]: v2 > v1
Out[5]: True

In [6]: v2 == v1
Out[6]: False
```

**Comment by csadorf** [comment]
Should you want to go with an automated solution after all (that would be my recommendation), I've had very good experience with [dunamai](https://pypi.org/project/dunamai/) and [aiialab-launch](https://github.com/aiidalab/aiidalab-launch/blob/main/aiidalab_launch/version.py) recently.

Even if you decide to go with a manual approach, might be worthwhile to compare against to either way.

**Comment by sphuber** [comment]
Thanks for the link @csadorf . I think this solution is too powerful for what we need. This seems to be really strong if you want to have a version that is granular on the level of the commit. But for AiiDA, this is even probably counter-productive as it will interfere with the caching mechanism in a probably undesirable and unexpected way (the version of `aiida-core` is included in the hash).

**Comment by csadorf** [comment]
> Thanks for the link @csadorf . I think this solution is too powerful for what we need. This seems to be really strong if you want to have a version that is granular on the level of the commit. But for AiiDA, this is even probably counter-productive as it will interfere with the caching mechanism in a probably undesirable and unexpected way (the version of `aiida-core` is included in the hash).

That's fair, but I would like to point out that you have control over how granular you want it to be. If you use the `no_metadata` option it will not include the commit. If that's not enough, you have full control over the pattern that is to be used. So if all you need is to add the post tag, that would be possible. Either way it's up to you, but I would suggest to at least have a look at the way that dunamai is implemented since they probably put some thought into how to apply these patterns consistenly.
